designs/0018-tuples.md

```diff
@@ -0,0 +1,395 @@
+- Feature Name: Tuples
+- Start Date: 2020-04-23
+- RFC PR:
+- Stan Issue: https://github.com/stan-dev/stan/issues/2431
+
+# Summary
+[summary]: #summary
+
+This proposal is to add tuples to the Stan language. Tuples are _static_,
+heterogeneous containers, meaning that they have a known, fixed size in the
+program and can contain elements which are of different types and sizes. Tuples
+are common in many languages from Python to OCaml and C++. Formally, a tuple is
+a product type of its contained types.
+
+# Motivation
+[motivation]: #motivation
+
+Tuples support several important use cases.
+
+One common use for tuples is to support _multiple return_, allowing functions to
+return more than one type of value simultaneously. For example, a function could
+be defined which returns both a matrix and some real-valued property of this
+matrix. In the math library, this can be used for cases such as singular value
+decomposition. Currently, two different functions must be called to receive the
+two portions of the decomposition, and this duplicates computation.
+
+Some forms of data are best represented by a mix of different types or different
+sizes of the same type; currently it is not possible in Stan to package these
+together.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A tuple is the generalization of the idea of a pair to any number of elements.
+In Stan, tuples allow packaging of different types together into one value.
+
+## Declaration
+
+A tuple variable is declared as follows
+
+```
+tuple(T1, ..., TN) x;
+```
+
+where `T1` through `TN` are themselves type specifications. Tuples are
+recursive - they can contain other tuples.
+
+A pair containing one integer and one real value could be declared
+
+```
+tuple(int, real) p;
+```
+
+These types can be any Stan type which would be valid in that block/context. For
+example, one can define a parameter which is a pair of an array of real values and a
+simplex like so
+
+```
+tuple(array[5] real, simplex[10]) param;
+```
+
+Tuples can be used as function arguments and return types.
+
+## Tuple expressions
+
+The expression for creating tuples will be `(a, b, c)`.  This will create a tuple
+of type `tuple(typeof(a), typeof(b), typeof(c))`, each subject to the normal
+promotion rules.
+
+This is inspired by the syntax currently used for arrays (`{value, value2, ...}`),
+and for row vector expressions (`[value1, value2, ...]`) with the notable
+difference that tuples allow mixes of types among their different values.
+
+For example, `(1.5, {2.3, 4})` is a tuple with type `tuple(real, array[2] real)`.
+
+## Tuple accessors
+
+A key difference between tuples and other containers in Stan is that they have a
+fixed size which is known during compilation. As such, they are accessed using
+indexing which is also known during compilation. This is done using `.` followed
+by a integer literal after the tuple. This static indexing is also known as
+projecting.
```

Comment:
It would be helpful to mention accessors in the summary in order to answer the question of why must they be known, fixed size.